src/enveloping/FeeEstimator.ts

```diff
@@ -0,0 +1,431 @@
+// @ts-ignore
+import { DataFrame } from 'dataframe-js'
+import { configureServer, ServerConfigParams } from '../relayserver/ServerConfigParams'
+import { BlockTransactionObject } from 'web3-eth'
+import { HttpProvider, IpcProvider, Transaction, WebsocketProvider } from 'web3-core'
+import Web3 from 'web3'
+import Timeout = NodeJS.Timeout
+
+const BLOCK_TIME = 20
+
+/**
+ * These are the minimum needed % of blocks accepting gas prices. For example, a gas price that has been accepted at least on 30% of
+ * the recent mined blocks will be considered for safeLow gas price because it will take longer to be included. These values can be modified.
+ * @SAFELOW will be mined in < 30m
+ * @STANDARD will be mined in < 5m
+ * @FAST will be mined in < 2m
+ */
+const SAFELOW = 30
+const STANDARD = 60
+const FAST = 90
+
+/**
+ * The amount of last mined blocks to take as reference for analysis
+ */
+const BLOCKS_TO_ANALYZE = 100
+
+export type Web3Provider =
+  | HttpProvider
+  | IpcProvider
+  | WebsocketProvider
+
+export interface FeesTable {
+  safeLow: number
+  standard: number
+  fast: number
+  fastest: number
+  blockTime: number
+  blockNum: number
+}
+
+/**
+ * Monitors the blockchain and estimates the recommended gas prices for different tolerated delays
+ */
+export class FeeEstimator {
+  blockData: DataFrame
+  config: ServerConfigParams
+  currentBlock: number
+  initialized: Boolean
+  feesTable?: FeesTable
+  readonly web3: Web3
+  worker?: Timeout
+  workerSemaphore: Boolean
```

Comment:
we could probably separate some of these into a sort of state, not all of these attributes are at the same categorical level